compiler/qsc_frontend/src/resolve.rs

```diff
@@ -984,106 +1089,326 @@ fn decl_is_intrinsic(decl: &ast::CallableDecl) -> bool {
+///
+/// # Parameters
+///
+/// * `kind` - The [`NameKind`] of the name
+/// * `globals` - The global scope to resolve the name against.
+/// * `provided_symbol_name` - The symbol name to resolve.
+/// * `provided_namespace_name` - The namespace of the symbol, if any.
+/// * `vars` - A mutable reference to a boolean indicating whether to allow local variables in closures in parent scopes.
+/// * `scope` - The scope to check for resolutions.
+fn check_scoped_resolutions(
+    kind: NameKind,
+    globals: &GlobalScope,
+    provided_symbol_name: &Ident,
+    provided_namespace_name: &Option<Idents>,
+    vars: &mut bool,
+    scope: &Scope,
+) -> Option<Result<Res, Error>> {
+    if provided_namespace_name.is_none() {
+        if let Some(res) =
+            resolve_scope_locals(kind, globals, scope, *vars, &provided_symbol_name.name)
+        {
+            // Local declarations shadow everything.
+            return Some(Ok(res));
         }
     }
+    let aliases = scope
+        .opens
+        .iter()
+        .map(|(alias, opens)| {
+            (
+                alias.clone(),
+                opens.iter().map(|x| (x.namespace, x)).collect(),
+            )
+        })
+        .collect::<FxHashMap<_, _>>();
+
+    let explicit_open_candidates = find_symbol_in_namespaces(
+        kind,
+        globals,
+        provided_namespace_name,
+        provided_symbol_name,
+        scope
+            .opens
+            .iter()
+            .flat_map(|(_, open)| open)
+            .map(|open @ Open { namespace, .. }| (*namespace, open)),
+        &aliases,
+    );
+    match explicit_open_candidates.len() {
+        1 => {
+            return Some(Ok(single(explicit_open_candidates.into_keys())
+                .expect("we asserted on the length, so this is infallible")))
+        }
+        len if len > 1 => {
+            return Some(ambiguous_symbol_error(
+                globals,
+                provided_symbol_name,
+                explicit_open_candidates
+                    .into_iter()
+                    .map(|(a, b)| (a, b.clone()))
+                    .collect(),
+            ))
+        }
+        _ => (),
+    }
+    if scope.kind == ScopeKind::Callable {
+        // Since local callables are not closures, hide local variables in parent scopes.
+        *vars = false;
+    }
+    None
+}
 
-    if candidates.len() > 1 {
-        // If there are multiple candidates, remove unimplemented items. This allows resolution to
-        // succeed in cases where both an older, unimplemented API and newer, implemented API with the
-        // same name are both in scope without forcing the user to fully qualify the name.
-        let mut removals = Vec::new();
-        for res in candidates.keys() {
-            if let Res::Item(_, ItemStatus::Unimplemented) = res {
-                removals.push(*res);
+/// This function always returns an `Err` variant of `Result`. The error is of type `Error::Ambiguous` and contains
```

Comment:
It's a bit awkward to have a function that returns Result when it is intended to always return an Error. Since it is only ever called in one place, it might be better to have this function return `Error::Ambiguous` without the Result::Err wrapper, and have that wrapping take place at the call site.